token-lending/program/src/state/reserve.rs

```diff
@@ -92,6 +92,41 @@ impl Reserve {
         }
     }
 
+    /// Calculate the minimum collateral for an obligation
+    pub fn minimum_collateral(
+        &self,
+        obligation: &Obligation,
+        liquidity_token_mint: &Pubkey,
+        token_converter: impl TokenConverter,
+    ) -> Result<u64, ProgramError> {
+        Self::_minimum_collateral(
+            obligation,
+            liquidity_token_mint,
+            self.collateral_exchange_rate()?,
+            &self.config,
+            token_converter,
+        )
+    }
+
+    fn _minimum_collateral(
+        obligation: &Obligation,
+        liquidity_token_mint: &Pubkey,
+        collateral_exchange_rate: CollateralExchangeRate,
+        collateral_reserve_config: &ReserveConfig,
+        mut token_converter: impl TokenConverter,
+    ) -> Result<u64, ProgramError> {
+        let borrow_token_price = token_converter.best_price(liquidity_token_mint)?;
+        let loan_to_value =
+            obligation.loan_to_value(collateral_exchange_rate, borrow_token_price)?;
+        let reserve_loan_to_value =
+            Decimal::from_percent(collateral_reserve_config.loan_to_value_ratio);
+        let loan_to_value_ratio = loan_to_value.try_div(reserve_loan_to_value)?;
+
+        Decimal::from(obligation.deposited_collateral_tokens)
+            .try_mul(loan_to_value_ratio)?
+            .try_ceil_u64()
```

Comment:
This is overly complex. You can get by with `borrow_amount / reserve_loan_to_value` to calculate `minimum_collateral`. Can you simplify and change the signature of `minimum_collateral` to accept a borrow amount instead of an obligation?

Reply:
I'm wondering in that case if this function is even needed. Wouldn't this do almost the same thing?

```rust
let required_collateral = withdraw_reserve.required_collateral_for_borrow(
    obligation.borrowed_liquidity_wads.try_ceil_u64()?,
    &borrow_reserve.liquidity.mint_pubkey,
    trade_simulator,
)?;
```

The main difference is then that `minimum_collateral` uses `token_converter.best_price` and `required_collateral_for_borrow` uses `token_converter.convert` per https://github.com/solana-labs/solana-program-library/pull/1229#discussion_r575806271

But is this distinction worth having a separate function that's so similar?

Reply:
Yeah, it's debatable. The current thought is that the token converter method gives you an impact price which is more resistant to manipulation which prevents people from taking out borrowing more than they should. On the other hand, the health of a loan is judged on the best price because we want to make liquidations fair. Otherwise, if we also used impact price for liquidations, the size of a loan would make an obligation more or less vulnerable to liquidation which didn't seem fair. More details here: https://github.com/solana-labs/solana-program-library/pull/1119

Reply:
Got it, that all makes sense. In this case, since it doesn't affect liquidation, my thinking is that `withdraw_obligation_collateral` should prevent you from withdrawing more collateral than you should based on the impact price, because if you can borrow and then manipulate the best price, you can effectively take out more borrowing than you should after the fact.

Reply:
Removed in dcdcf9e